### PR TITLE
Structs in Moya.

### PR DESCRIPTION
**Pull request by sunshinejr**
As per issue #408, I'm sharing with you the solution I've come up with to the topic, which is structs in Moya. The general idea was a perfect start by @Matthijn followed up by a really good direction by @colinta 

## Motivation
We would like to keep existing enum setup and additionally support structs based setup for bigger projects or for any other reason. Specifically, what should be possible with this PR is that we can still use old configuration of enums and it won't break, but additionally we can use Moya with structs configuration, for instance:

```swift
GitHubStructProvider.request(.UserProfile(UserProfileResource(name: name))) { result in
            
}
```

## Implementation
My idea was to create protocol that will be a default protocol for enum when using it with wrapped structs, I called it `ServiceType`. Then make `TargetType` conform to it protocol so we can reuse code for both implementations.

## Issues

There were few issues that occurred during the implementation. Changes I've made created some problems with generic's type inferring so I had to move Defaults (`DefaultEndpointMapping`, `DefaultRequestMapping`, `DefaultAlamofireManager` and stubbing defaults) to another non-generic class and I called it `MoyaDefaults`, which is a **breaking change** in case someone used them via `MoyaProvider.DefaultAlamofireManager` for instance. Couldn't find a way for not-creating another struct without many changes in code so if you have an idea I would love to hear it. 

## Demo
I've also added an example of usage to the current demo. As per our conversation I also believe we could move it to second demo. In short we can now define our struct like the one below:

```swift
public enum GitHubStruct: ServiceType {
    case Zen(ZenResource)
    case UserProfile(UserProfileResource)
}

public struct ZenResource: TargetType {
    public let path = "/zen"
    public let baseURL = NSURL(string: "https://api.github.com")!
    public let method: Moya.Method = .GET
    public let parameters: [String: AnyObject]? = nil
    public let sampleData = "Half measures are as bad as nothing at all.".dataUsingEncoding(NSUTF8StringEncoding)!
}
```

## Conclusion
This is really a first step and we could improve it I believe, so I'm curious to hear what you think about it.

Cheers 🎉

### CODE REVIEW

**Comment by Matthijn** [comment]
I must say I like this solution. Haven't checked into this a while. But this looks promising.

File: Demo/Tests/MoyaProviderSpec.swift [hunk 1]
```diff
@@ -7,7 +7,7 @@ class MoyaProviderSpec: QuickSpec {
     override func spec() {
         var provider: MoyaProvider<GitHub>!
         beforeEach {
-            provider = MoyaProvider<GitHub>(stubClosure: MoyaProvider.ImmediatelyStub)
+            provider = MoyaProvider<GitHub>(stubClosure: MoyaDefaults.ImmediatelyStub)
```
Comment (colinta) [inline Demo/Tests/MoyaProviderSpec.swift]: I'm really liking this PR!  However, I'm gonna ask for a few changes to go into a separate PR, like the change from `MoyaProvider` to `MoyaDefaults`.  This is a breaking change, and I'm hoping we can support structs without any breaking changes.

Comment (sunshinejr) [inline Demo/Tests/MoyaProviderSpec.swift]: Thanks 🎉 Sure, will do this today. ✌️

File: Source/Moya.swift [hunk 2]
```diff
@@ -65,12 +87,16 @@ public class MoyaProvider<Target: TargetType> {
...
     
     /// Returns an Endpoint based on the token, method, and parameters by invoking the endpointsClosure.
-    public func endpoint(token: Target) -> Endpoint<Target> {
+    public func endpoint(token: TargetType) -> Endpoint<TargetType> {
         return endpointClosure(token)
     }
     
     /// Designated request-making method. Returns a Cancellable token to cancel the request later.
     public func request(target: Target, completion: Moya.Completion) -> Cancellable {
+        guard let target = target.resource else {
```
Comment (ashfurrow) [inline Source/Moya.swift]: Could you expand on the use of `.resource` here, I'm not sure I understand it or `ServiceType` but I'd like to know more!

File: Demo/Demo/GitHubStructAPI.swift [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    }
+    public var parameters: [String: AnyObject]? {
+        return nil
+    }
+    public var sampleData: NSData {
+        return "Half measures are as bad as nothing at all.".dataUsingEncoding(NSUTF8StringEncoding)!
+    }
+    
+    public let name: String
+}
```
Comment (ashfurrow) [inline Demo/Demo/GitHubStructAPI.swift]: Are these two different to demonstrate flexibility? I like flexibility, and this looks cool, but I think using the syntax from `ZenResource` for all of these properties (except `path`) would help encourage users to follow that pattern where they can. I like opinionated code :wink:

File: Source/Moya.swift (hunk 2 above)
Comment (sunshinejr) [inline Source/Moya.swift]: Of course! So the general idea of `ServiceType` is that this is the core of **StructsAPI**. Every enum that wants to use Moya with structs would have to conform to `ServiceType`. We then believe that when you conform to that protocol, you know what are you doing (I mean that is the risk I was talking about on hangout). So as a follow-up, I continue to believe the user and I try to fetch the associated object as it is a Resource (TargetType). But! I've made `TargetType` to conform to `ServiceType` as a default. So if you are using **EnumAPI**, it basically means that we don't need to look for associated object, we are the associated object, we are a TargetType (Resource). 

I don't know if that seems good to you, but let me know, I will try harder express my mind 🤕

File: Demo/Demo/GitHubStructAPI.swift (hunk 3 above)
Comment (sunshinejr) [inline Demo/Demo/GitHubStructAPI.swift]: Yeah, I've added this as a second option with using properties of a struct as a parameters. Sure, will do!

File: Source/Moya.swift (hunk 2 above)
Comment (ashfurrow) [inline Source/Moya.swift]: OK, that makes sense, thanks. I guess I was thinking of something like this:

```swift
enum StructTargetType: TargetType {
  public var path: String {
    switch self {
      case .Struct(let t): return t.path
    }
  }
  // and so on
}
```

So that the struct-based approach totally piggy-backed on the existing enum structure, instead of having the existing enum structure conform to something new. 

That was totally just _my_ interpretation, it may not be possible. You've played with this more than I have :smile:

Comment (colinta) [inline Source/Moya.swift]: What Ash describes here is also what I was imagining - start with what we have, instead of adding anything new, just update the examples to show that *struct based endpoints is already possible*.

**Comment by ashfurrow** [comment]
Going to close this in lieu of #427.
